File: baileys/src/wa/phone.ts
export class InvalidPhoneNumberError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'InvalidPhoneNumberError'
  }
}

const MIN_DIGITS = 8
const MAX_DIGITS = 15

export function normalisePhoneNumber(input: string): string {
  const raw = input.trim()
  const digits = raw.replace(/\D/g, '')

  if (digits.length === 0) {
    throw new InvalidPhoneNumberError(`"${raw}" contains no digits`)
  }

  if (digits.startsWith('00')) {
    throw new InvalidPhoneNumberError(
      `"${raw}" starts with the 00 international prefix. Drop it and use the country code ` +
        `alone, e.g. 00628123456789 becomes 628123456789.`,
    )
  }

  if (digits.startsWith('0')) {
    throw new InvalidPhoneNumberError(
      `"${raw}" looks like a national number. WhatsApp needs the country code and no ` +
        `leading zero, e.g. Indonesian 081234567890 becomes 6281234567890.`,
    )
  }

  if (digits.length < MIN_DIGITS || digits.length > MAX_DIGITS) {
    throw new InvalidPhoneNumberError(
      `"${raw}" has ${digits.length} digits; a full international number has ` +
        `${MIN_DIGITS}–${MAX_DIGITS}.`,
    )
  }

  return digits
}

export function toUserJid(input: string): string {
  return `${normalisePhoneNumber(input)}@s.whatsapp.net`
}

File: go-consumer/wa-shared/wa/phone.go
package wa

import (
	"fmt"
	"strings"
)

const (
	MinPhoneDigits = 8
	MaxPhoneDigits = 15

	DefaultCountryCode = "62"
)

func NormalisePhone(input string) (string, error) {
	raw := strings.TrimSpace(input)

	var b strings.Builder
	b.Grow(len(raw))
	for _, r := range raw {
		if r >= '0' && r <= '9' {
			b.WriteRune(r)
		}
	}
	digits := b.String()

	switch {
	case digits == "":
		return "", fmt.Errorf("%w: %q contains no digits", ErrInvalidPayload, raw)

	case strings.HasPrefix(digits, "00"):
		return "", fmt.Errorf("%w: %q starts with the 00 international prefix; drop it and send the country code alone, e.g. 6287713848500",
			ErrInvalidPayload, raw)

	// A leading zero means the national format, and the only country this API serves
	// nationally is Indonesia. A number that is already international is never
	// reinterpreted — see openapi.yaml PhoneNumber.
	case strings.HasPrefix(digits, "0"):
		digits = DefaultCountryCode + digits[1:]
	}

	if len(digits) < MinPhoneDigits || len(digits) > MaxPhoneDigits {
		return "", fmt.Errorf("%w: %q has %d digits; a full international number has %d–%d",
			ErrInvalidPayload, raw, len(digits), MinPhoneDigits, MaxPhoneDigits)
	}

	return digits, nil
}
